File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
  var answer = [];
  const map = {};

  for (const term of terms) {
    const splited = term.split(" ");
    map[splited[0]] = parseInt(splited[1]);
  }

  let idx = 1;
  for (const privarcy of privacies) {
    const splited = privarcy.split(" ");
    const parts = splited[0].split(".");

    let year = parseInt(parts[0], 10);
    let month = parseInt(parts[1], 10);
    let day = parseInt(parts[2], 10);

    year += Math.floor((month + map[splited[1]] - 1) / 12);
    month = (month + map[splited[1]] - 1) % 12 + 1;

    if (month < 10) {
      month = '0' + month;
    }
    if (day < 10) {
      day = '0' + day
    }
    const dateString = year + '.' + month + '.' + day;
    
    if (dateString.toString() <= today) {
      answer.push(idx);
    }
    idx++;
  }
  console.log(answer);
  return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
from typing import Dict, List

DAYS_PER_MONTH = 28
MONTHS_PER_YEAR = 12

def to_days(date: str) -> int:
    year, month, day = map(int, date.split('.'))
    return (year * MONTHS_PER_YEAR + month) * DAYS_PER_MONTH + day

def solution(today: str, terms: List[str], privacies: List[str]) -> List[int]:
    expiration: Dict[str, int] = {}
    for term in terms:
        key, months = term.split()
        expiration[key] = int(months)
    today_days = to_days(today)
    expired_indices = []
    for idx, privacy in enumerate(privacies, start=1):
        date_str, term_key = privacy.split()
        start_days = to_days(date_str)
        expiry_days = start_days + expiration[term_key] * DAYS_PER_MONTH - 1
        if today_days > expiry_days:
            expired_indices.append(idx)
    return expired_indices
